test/helpers/position_assertions.cc

```diff
@@ -1187,44 +1205,58 @@ void SymbolSearchAssertion::checkAllForQuery(std::string query,
         auto &location = symbol->location;
         auto line = fmt::format("{}:{:{}}", location->uri, location->range->start->line, lineWidth);
         symbolsByLine[line].push_back(symbol);
+        assertionsByLine[line]; // create entry in assertionsByLine, if none exists yet.
     }
 
-    bool success = true;
-    // Show any expected symbols that weren't returned by LSP
+    int numUnfulfilled = 0;
+    int numUnexpected = 0;
+    // Look line-by-line for unmatched symbols and assertions
     for (auto entry : assertionsByLine) {
         auto lineInfo = entry.first;
-        auto expected = entry.second;
-        auto &actual = symbolsByLine[lineInfo];
-        if (expected.size() == actual.size()) {
-            continue;
+        auto assertionsOnLine = entry.second;
+        auto &symbolsOnLine = symbolsByLine[lineInfo];
+
+        // Report assertions with no matching symbol
+        for (auto &assertion : assertionsOnLine) {
+            bool foundMatchingSymbol = false;
+            for (auto &symbol : symbolsOnLine) {
+                if (assertion->matches(uriPrefix, symbol)) {
+                    foundMatchingSymbol = true;
+                    break;
+                }
+            }
+            if (foundMatchingSymbol) {
+                continue;
+            }
+            numUnfulfilled++;
+            addFailureAtLocationWithSource(assertion->getLocation(uriPrefix), sourceFileContents, uriPrefix,
+                                           fmt::format("Unfulfilled assertion: {}", assertion->toString()));
         }
-        success = false;
-        auto assertion = expected[0];
-        auto location = assertion->getLocation(uriPrefix);
-        auto sourceLine = getLine(sourceFileContents, uriPrefix, *location.get());
-        ADD_FAILURE_AT(location->uri.c_str(), assertion->range->start->line + 1)
-            << fmt::format("WorkspaceSymbol query `{}` did not return a reference to:\n{}\n", query,
-                           prettyPrintRangeComment(sourceLine, *assertion->range, ""));
-    }
-
-    // Show any symbols returned by LSP that weren't asserted
-    for (auto entry : symbolsByLine) {
-        auto lineInfo = entry.first;
-        auto actual = entry.second;
-        auto &expected = assertionsByLine[lineInfo];
-        if (expected.size() == actual.size()) {
-            continue;
+        // Report symbols with no matching assertion
```

Comment:
Would it make sense to track if an assertion was matched with a mutable boolean on the assertion object? Then we can update that property in the loop above this one, and simplify the logic of this loop. (I'm thinking more about LOC here than speed.)

Reply:
If you're suggesting that it would simplify the code, I don't think that's true...IIUC, a correct solution requires backtracking and is isomorphic to [a variant of the Stable Marriage problem](https://en.wikipedia.org/wiki/Stable_marriage_with_indifference).  I would be a little concerned that adding a boolean without also supporting backtracking could cause the assertions to (seemingly non-deterministically) pass/fail depending on the order of symbols returned by the LSP.  Talk in person?

Reply:
Yeah, let's